### ISSUE TITLE
Questions about CTR problem with TFRanking

### ISSUE THREAD

**Issue by AlexanderJLiu** [tensorflow/ranking#207]
Hi, all:

I want to use TFRanking for my CTR problem, and the data for CTR model comes from the implicit click feedback logs.

In my case, CTR training/validation data is tuples of **(user features, item features, context features, click)** where user features contain age, gender etc, item features contain some item side information and context features(quite different from the context feature in TFRanking) contain time(e.g. week), area etc. All the features are represented by number index (generated from vocabulary).

According to this [comment](https://github.com/tensorflow/ranking/issues/110#issuecomment-527616929) about recsys with TFRanking, I retrofit my data format to adapt TFRanking model. Steps are shown as follow:

1. I use all the `user and context feature` as the `context feature` of TFRanking model.
2. I use all the `item features` as the `example feature` of TFRanking model.
3. After retrofitting, the `ELWC` data format looks like:

   ```protobuf
    context {
      features {
        feature {
          key: "query_tokens"
          value {
            int64_list {
              value: 12 //supposed to be user feature
              value: 39
              value: 44 //supposed to be context feature
            }
          }
        }
      }
    }
    examples {
      features {
        feature {
          key: "document_tokens"
          value {
            int64_list {
              value: 56 // item feature
              value: 67
              value: 69
              value: 89
            }
          }
        }
        feature {
          key: "relevance"
          value {
            int64_list {
              value: 1
            }
          }
        }
      }
    }
   ```

I generate the tfrecord data follow this [comment](https://github.com/tensorflow/ranking/issues/196#issuecomment-631236916) using the data format above. Note that due to the original data format, The `ELWC` query(sample) data stored in tfrecord files `only have one example in the example set` which is relevant or irrelevant to the context.

I then train the TFRanking model with `SOFTMAX_LOSS`(list wise loss), and find the loss always equals to 0. If I provide one more example in the example set, the loss will change for every batch, but why?

Training code could be found in this [colab](https://colab.research.google.com/drive/1VT7NIqMlwuymIQMiRpMuDoVU2sAfLWbQ?usp=sharing)

**Here are my questions:**

1. Is it the right way to create `context feature` and `example feature` for TFR model according to the data retrofit part?
2. How does the loss be computed across training? Why zero loss occur in training process?
3. Do we have to provide more than one example for every query(sample)? In other words, put all the examples `relevant/irrelevant` to the same context together for a single query(sample)?

Any reply will be very very appreciated!!!

Thanks~

**Comment by eggie5** [comment]
> I use all the user and context feature as the context feature of TFRanking model.
> 
> I use all the item features as the example feature of TFRanking model.

These points are correct. 

For context features you have:
* age
* gender 
* time
* area

However, when you showed your example feature proto for context it only has a `query_tokens` feature, which wasn't in your original list of features:

```
 context {
   features {
     feature {
       key: "query_tokens"
       value {
         int64_list {
           value: 12 //supposed to be user feature
           value: 39
           value: 44 //supposed to be context feature
         }
       }
     }
   }
 }
```

**Comment by AlexanderJLiu** [comment]
> However, when you showed your example feature proto for context it only has a query_tokens feature, which wasn't in your original list of features

@eggie5 My bad, I made it ambiguous. I'll explain it in more detail. 

As you say, I have age, gender... in context features, in general, I should use separate vocabulary for each context feature and the ELWC data  format should like this:

```protobuf
context {
  features {
    feature {
      key: "age"
      value {
        int64_list {
          value: 3
        }
      }
    }
    feature {
      key: "gender"
      value {
        int64_list {
          value: 1
        }
      }
    }
    feature {
      key: "week"
      value {
        int64_list {
          value: 5
        }
      }
    }
    ...
  }
}
```

But for simplify, I just combine all the features(age, gender...) to one `query_token` feature,  and all the features share the same vocabulary file, thus their index representation should be global and not private. That's why the context feature look like this:

```protobuf
 context {
   features {
     feature {
       key: "query_tokens"
       value {
         int64_list {
           value: 12 //supposed to be user feature
           value: 39
           value: 44 //supposed to be context feature
         }
       }
     }
   }
 }
```

Finally in code, I can just use one `embedding column` for the context feature(cause all the features share the same embedding matrix).

In all my CTR model like `DeepFM`, I just handle all the features using the same vocabulary globally, so I move it to TFRanking. 

I know that this way may be not suitable for feature analysis, like which feature would be more important or which would be useless for the model. But If I ignore that, it should be fine for training the model with the data like that.

I'm not pretty sure, please correct me if I'm wrong.

Thanks very much~

**Comment by AlexanderJLiu** [comment]
According to the list-wise loss function defined in TFRanking [paper](https://arxiv.org/pdf/1812.00073.pdf) and the metric explanation [here](https://medium.com/swlh/rank-aware-recsys-evaluation-metrics-5191bba16832), I may find the reason why `zero loss` occur. 

The loss is similar with `multi-classification task`, the difference is classes in multi-classification task will never be `only one`, But in LTR, if we only have one example in example set(i.e. one class), the loss will always be equal to `zero`. And the metric like ndcg will always be equal to `1`.  

Is my analysis correct? If so, we have to put all the examples relevant/irrelevant to the same context together, is it? And does it mean the context with only one example gives useless information for the TFR model(or how to handle this situation)?

Thanks~

**Comment by xuanhuiwang** [comment]
@AlexanderJLiu, you are right for pairwise or listwise loss: you need to have >= 2 documents with different labels in an ELWC. You can still use the pointwise loss, sigmoid_cross_entropy_loss, for you CTR problem -- which is the same as the regression setting though.

**Comment by xuanhuiwang** [comment]
Feel free to open if you have more questions.

**Comment by zahraegh** [comment]
@xuanhuiwang Do we need to have >=2  in the examples list with various labels in both train and test sets? or only the train set?
